### PR TITLE
Support negative indexes in tuple slices

### PR DESCRIPTION
**Pull request by axitkhurana**
#886

### CODE REVIEW

File: mypy/checkexpr.py [hunk 1]
```diff
@@ -1136,26 +1136,62 @@ def visit_index_expr_helper(self, e: IndexExpr) -> Type:
...
     def visit_tuple_slice_helper(self, left_type: TupleType, slic: SliceExpr):
-        begin = 0
-        end = len(left_type.items)
-        stride = 1
+        begin = None  # type: int
+        end = None  # type: int
+        stride = None  # type:int
+
         if slic.begin_index:
+            ok = False
```
Comment (axitkhurana) [inline mypy/checkexpr.py]: Feel like I'm repeating myself for begin_index, end_index and stride. Wondering if it's worth moving the logic to an _function

File: mypy/test/data/check-tuples.test [hunk 2]
```diff
@@ -172,6 +173,7 @@ b = t1[1]
 a = t1[(0)]
 x = t3[0:3] # type (A, B, C)
 y = t3[0:5:2] # type (A, C, E)
+z = t3[:-1] # type (A, B, C, D)
```
Comment (gnprice) [inline mypy/test/data/check-tuples.test]: This should be `(A, B, C)`, right?

Comment (gnprice) [inline mypy/test/data/check-tuples.test]: Actually, I was confused by `z` above. Why was that line added?

File: mypy/checkexpr.py (hunk 1 above)
Comment (gnprice) [inline mypy/checkexpr.py]: Yeah, let's definitely factor out that repetition.

**Comment by axitkhurana** [comment]
Text for commit message:

Add support negative indexes in tuple slices

    * change defaults for start/end/stride to `None` to support negative indexes
    * refactor index value extraction from start/end/stride into a function
    * add test cases for negative index access and negative index tuple slices
